qtcodes/circuits/circ.py

```diff
@@ -16,60 +16,66 @@ class TopologicalRegister:
+            params (List[Dict[str,int]]):
+                 Contains a list of params such as d, where d is the number of
                  physical "data" qubits lining a row or column of the lattice.
             name (str):
                 Useful when combining multiple TopologicalQubits together.
                 Prepended to all registers.
 
         """
-        self.params = params if params else {"d": 3}
+        ctypes = [] if ctypes is None else ctypes
+        params = [] if params is None else params
+        self.params = []    
         self.name = name
-
         self.n = 0
-
         # == None is necessary, as "not circ" is true for circ=QuantumCircuit()
         self.circ = QuantumCircuit() if circ is None else circ
-
-        if ctype not in str2qtype:
+        self.tqubits: Dict[str, Dict[int, Type[Any]]] = {}
+        self.add_tqubits("data", ctypes, params)
+    
+    def add_tqubits(
+        self, 
+        sub_register: str, 
+        ctypes: Optional[List[str]] = None, 
+        params: Optional[List[Dict[str, int]]] = None
+    ) -> None:
+        """
+        Allows us to add logical qubits to TopologicalRegister
+        """
+        params = [] if params is None else params
+        ctypes = [] if ctypes is None else ctypes
+        if len(params) != len(ctypes):
             raise ValueError(
-                "Please choose a Topological Qubit type from: "
-                + str(list(str2qtype.keys()))
+                "Please match the number of params with the number of Topological Qubits added: Current number of params - " + str(len(params)) + ", Current number of Topological Qubits: " + str(len(ctypes)) 
             )
-        self.tqubit_type = str2qtype[ctype]
-        self.tqubits: Dict[str, Dict[int, Type[Any]]] = {}
-        for _ in range(num_tqubits):
-            self.add_tqubit("data")
-
-    def add_tqubit(self, sub_register: str) -> None:
-        if sub_register not in self.tqubits:
-            self.tqubits[sub_register] = {}
-        self.tqubits[sub_register][self.n] = self.tqubit_type(
-            params=self.params, name=self.name + "_" + str(self.n), circ=self.circ
-        )
-        self.n += 1
-
-    def add_tqubits(self, sub_register: str, num_tqubits: int) -> None:
-        for _ in range(num_tqubits):
-            self.add_tqubit(sub_register)
+        for i in range(len(ctypes)):
+            if ctypes[i] not in str2qtype:
+                raise ValueError(
+                    "Please choose a Topological Qubit type from: "
+                    + str(list(str2qtype.keys()))
+                )
+        for i in range(len(ctypes)):
+            if sub_register not in self.tqubits:
+                self.tqubits[sub_register] = {}
+            self.tqubits[sub_register][self.n] = str2qtype[ctypes[i]](
+                params=params[i], name=self.name + "_" + str(self.n), circ=self.circ
+            )
+            self.params.append(params[i])
+            self.n += 1
```

Comment:
This should be moved to another method with signature: `add_tqubit(self, sub_register, ctype: str, params: Dict[str, int])`.